channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+#[derive(Serialize, Debug, Clone)]
+pub struct SenderData {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub ua: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub addr: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub city: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub region: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub country: Option<String>,
+}
+
+impl Default for SenderData {
+    fn default() -> Self {
+        Self {
+            ua: None,
+            addr: None,
+            city: None,
+            region: None,
+            country: None,
+        }
+    }
+}
+
+// Parse the Accept-Language header to get the list of preferred languages.
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
+    let default_lang = String::from("en");
+    let mut lang_tree: BTreeMap<String, String> = BTreeMap::new();
+    let mut i = 0;
+    alheader.split(",").for_each(|l| {
+        if l.contains(";") {
+            let weight: Vec<&str> = l.split(";").collect();
+            let lang = weight[0].to_ascii_lowercase();
+            let pref = weight[1].to_ascii_lowercase();
+            lang_tree.insert(String::from(pref), String::from(lang));
+        } else {
+            lang_tree.insert(
+                format!("q=1.{:02}", i),
+                String::from(l.to_ascii_lowercase()),
+            );
+            i += 1;
+        }
+    });
+    let mut langs: Vec<String> = lang_tree.values().map(|l| l.to_owned()).collect();
+    langs.reverse();
+    langs.push(default_lang);
+    langs
+}
+
+// Return the element that most closely matches the preferred language.
+// This rounds up from the dialect if possible.
+fn get_preferred_language_element(
+    langs: &Vec<String>,
+    elements: BTreeMap<String, String>,
+) -> Option<String> {
+    for lang in langs.clone() {
+        if elements.contains_key(&lang) {
+            //unwrap is safe-ish because we've checked that the key exists.
+            return Some(elements.get(lang.as_str()).unwrap().to_string());
+        }
+        if lang.contains("-") {
+            let (lang, _) = lang.split_at(2);
+            if elements.contains_key(lang) {
+                return Some(elements.get(lang).unwrap().to_string());
+            }
+        }
+    }
+    None
+}
+
+// Set the sender meta information from the request headers.
+impl From<HttpRequest<WsChannelSessionState>> for SenderData {
+    fn from(req: HttpRequest<WsChannelSessionState>) -> Self {
+        let mut sender = SenderData::default();
+        let headers = req.headers();
+        //TODO: Get the default lang
+        let langs = match headers.get(http::header::ACCEPT_LANGUAGE) {
+            None => vec![String::from("en")],
+            Some(l) => preferred_language(l.to_str().unwrap_or("").to_owned()),
```

Comment:
Under what circumstances would `to_str` fail here?  Is it if the client submits invalid utf8?  If so, would it be better to fail the entire request on the assumption that something has gone horribly wrong?

Also, perhaps we should log the failure here in a similar manner to the handling of user-agent below.

Reply:
I'm not really sure how this would fail. Bad UTF-8 is an option, of course. The problem is that this is a From:: so it has to return itself, not an option. I'm going to instead switch from "en" default to "*" and let random ordering sort them out.